packages/headless/src/api/search/date/relative-date.ts

```diff
@@ -0,0 +1,167 @@
+  'past',
+  'now',
+  'future',
+];
+
+/**
+ * The unit to which to can set the date relative to.
+ */
+export type RelativeDateUnit =
+  | 'minute'
+  | 'hour'
+  | 'day'
+  | 'week'
+  | 'month'
+  | 'quarter'
+  | 'year';
+const validRelativeDateUnits: RelativeDateUnit[] = [
+  'minute',
+  'hour',
+  'day',
+  'week',
+  'month',
+  'quarter',
+  'year',
+];
+
+/**
+ * Allows to define a date relative to the current moment.
+ */
+export interface RelativeDate {
+  /**
+   * The relative period of time.
+   */
+  period: RelativeDatePeriod;
+  /**
+   * The unit of time. When `period` is `now`, does not have to be defined.
+   */
+  unit?: RelativeDateUnit;
+  /**
+   * The amount of the `unit` of time. When `period` is `now`, does not have to be defined.
+   */
+  amount?: number;
+  /**
+   * If `true`, the date will be returned unshifted. If `false`, the date will be adjusted to UTC time.
+   *
+   * @defaultValue `false`
+   */
+  useLocalTime?: boolean;
+}
+
+const buildRelativeDateDefinition = (period: RelativeDatePeriod) => {
+  const isNow = period === 'now';
+  return {
+    amount: new NumberValue({required: !isNow, min: 1}),
+    unit: new StringValue({
+      required: !isNow,
+      constrainTo: validRelativeDateUnits,
+    }),
+    period: new StringValue({
+      required: true,
+      constrainTo: validRelativeDatePeriods,
+    }),
+    useLocalTime: new BooleanValue(),
+  };
+};
+
+function validateRelativeDate(relativeDate: RelativeDate) {
+  new Schema(buildRelativeDateDefinition(relativeDate.period)).validate(
+    relativeDate
+  );
+}
+
+/**
+ * Formats a `RelativeDate` object to a parsable string value.
+ * @param relativeDate The `RelativeDate` object.
+ */
+export function formatRelativeDate(relativeDate: RelativeDate) {
+  validateRelativeDate(relativeDate);
+  const {period, amount, unit, useLocalTime} = relativeDate;
+  const utc = useLocalTime ? '' : 'utc';
+
+  switch (period) {
+    case 'past':
+    case 'future':
+      return `${period}${amount}${unit}${utc}`;
+    case 'now':
+      return `${period}${utc}`;
+  }
+}
+
+export function formatRelativeDateForSearchApi(date: string) {
+  try {
+    const relativeDate = parseRelativeDate(date);
+    const {period, amount, unit, useLocalTime} = relativeDate;
+    const dayjsDate = useLocalTime ? dayjs() : dayjs().utc();
+
+    switch (period) {
+      case 'past':
+        return formatDateForSearchApi(
+          dayjsDate.subtract(amount!, unit as QUnitType)
+        );
+      case 'future':
+        return formatDateForSearchApi(
+          dayjsDate.add(amount!, unit as QUnitType)
+        );
+      case 'now':
+        return formatDateForSearchApi(dayjsDate);
+    }
+  } catch (error) {
+    return formatDateForSearchApi(dayjs());
+  }
+}
+
+const relativeDateFormatRegexp = new RegExp(
+  `^((?<period>past|future)(?<amount>\\d+)(?<unit>${validRelativeDateUnits.join(
+    '|'
+  )})|(?<now>now))(?<utc>utc)?$`,
```

Comment:
@klagace-coveo, I wanted to get your opinion on this regex. Can it be abused?
The string is used in the url, so it can be modified to a different value before being shared via a link.

Suggestion: would it be safer and simpler to introduce a delimiter and not use a regex? e.g.

```
const start = 'past-1-year';
const [period, amount, units] = start.split('-');
```

Reply:
I don't get the issue with using a regex 😂 
The way its used, if you call `isRelativeDateFormat` which does `relativeDateFormatRegexp.test(date)` and it return false, if won't see the Url value as valid and move on. How is a split safer?

Reply:
Maybe there's something about regexes and XSS I need to be aware of...

Reply:
Regex security problem comes from two things normally:
* Trying to use regex to sanitize something that gets added to the DOM (as HTML) -> XSS
* Regex used server side to handle user input -> You can DDOS because an improperly crafted regex can create infinite loops

Both of these don't apply here...

Reply:
Looking at the regex again, I see that the dynamic portion is just the relativeDateUnits. There is no user input so it is safe!

I still think a delimiter approach would be simpler to maintain though,

Reply:
With a delimiter would still need to validate each parts separately, regex takes care of all in one

Reply:
I don't have an issue with a regex, but I do like the reduced complexity of a simple split with dashes.

Reply:
Let's go with a split then